File: transpile/templates/java/src/test/java/trld/jsonld/TestSuite.java
package trld.jsonld;

import java.util.*;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import static org.junit.jupiter.api.Assertions.*;

import static trld.jsonld.Base.CONTEXT;


public class TestSuite {

    static final String TESTSUITE_DIR = "file://" + System.getProperties().get("user.home") + "/repos/github/w3c/json-ld-api/tests/";
    static final String TESTSUITE_BASE_URL = "https://w3c.github.io/json-ld-api/tests/";

    static List<Arguments> testCaseProvider(String manifestFile) {
        List<Arguments> cases = new ArrayList<>();
        Map manifest = (Map) Common.loadJson(TESTSUITE_DIR + manifestFile);
        for (Map tc : (List<Map>) manifest.get("sequence")) {
            if (!((List) tc.get("@type")).contains("jld:PositiveEvaluationTest"))
                continue; // TODO: add handling of negative tests!

            cases.add(Arguments.of(((String) tc.get("@id")),
                                   ((String) tc.get("purpose")), tc));
        }
        return cases;
    }

    static List<Arguments> expandTestCases() {
        return testCaseProvider("expand-manifest.jsonld");
    }

    static List<Arguments> compactTestCases() {
        return testCaseProvider("compact-manifest.jsonld");
    }

    @ParameterizedTest(name = "Expand {0} - {1}")
    @MethodSource("expandTestCases")
    void expandTestSuite(String id, String purpose, Map tc) {
        String src = TESTSUITE_DIR + (String) tc.get("input");
        Map<String, Object> options = (Map) tc.getOrDefault("option", new HashMap<>());
        String baseUri = (String) options.getOrDefault("base", src);
        boolean ordered = true;

        String expandContext = (String) options.get("expandContext");
        if (expandContext != null) {
            expandContext = TESTSUITE_DIR + expandContext;
        }

        Object sourceData = Common.loadJson(src);

        String expectedSrc = TESTSUITE_DIR + (String) tc.get("expect");
        Object expectedData = Common.loadJson(expectedSrc);

        Object resultData = Expansion.expand(sourceData, baseUri, expandContext, ordered);

        assertEquals(expectedData, resultData);
    }

    @ParameterizedTest(name = "Compact {0} - {1}")
    @MethodSource("compactTestCases")
    void compactTestSuite(String id, String purpose, Map tc) {
        String src = TESTSUITE_DIR + (String) tc.get("input");
        String url = TESTSUITE_BASE_URL + (String) tc.get("input");
        Map<String, Object> options = (Map) tc.getOrDefault("option", new HashMap<>());
        String baseUri = (String) options.getOrDefault("base", url);
        boolean compactArrays = (Boolean) options.getOrDefault("compactArrays", true);
        boolean ordered = true;

        Object sourceData = Common.loadJson(src);
        sourceData = Expansion.expand(sourceData, baseUri, null, ordered);
        String contextSrc = TESTSUITE_DIR + (String) tc.get("context");
        Object contextData = Common.loadJson(contextSrc);

        String expectedSrc = TESTSUITE_DIR + (String) tc.get("expect");
        Object expectedData = Common.loadJson(expectedSrc);
        if (expectedData instanceof Map) {
            ((Map) expectedData).remove(CONTEXT);
        }

        Object resultData = Compaction.compact(contextData, sourceData, baseUri, compactArrays, ordered);

        assertNotNull(expectedData);
        assertEquals(expectedData, resultData);
    }
}

File: transpile/templates/js/test/test_suite.js
import tape from 'tape'
import os from 'os'

import { anyDocumentLoader, setDocumentLoader } from '../lib/jsonld/docloader.js'
import * as context from '../lib/jsonld/context.js'
import { JSONLD10, CONTEXT, ID, TYPE } from '../lib/jsonld/base.js'
import { expand } from '../lib/jsonld/expansion.js'
import { compact } from '../lib/jsonld/compaction.js'
import { flatten } from '../lib/jsonld/flattening.js'

const TESTSUITE_DIR = `file://${process.env.TRLD_JSONLD_TESTDIR}/`
const TESTSUITE_BASE_URL = 'https://w3c.github.io/json-ld-api/tests/'

function localTestsuiteLoader(url) {
  return anyDocumentLoader(url.replace(TESTSUITE_BASE_URL, TESTSUITE_DIR))
}

setDocumentLoader(localTestsuiteLoader)

function loadJson(url) {
  let data = localTestsuiteLoader(url).document
  return data
}

function testCaseRunner(category) {
  let manifestFile = `${TESTSUITE_DIR}${category}-manifest.jsonld`
  let manifest = loadJson(manifestFile);
  for (let tc of manifest.sequence) {
    if (tc[TYPE].indexOf('jld:PositiveEvaluationTest') === -1) {
      continue
    }
    let label = `${category} ${tc[ID]} - ${tc.purpose}`
    let options = tc.option || {}
    if (options.specVersion === JSONLD10) {
        continue
    }

    tape.test(label, t => {
      let src = TESTSUITE_DIR + tc.input
      let url = TESTSUITE_BASE_URL + tc.input
      let baseUri = options.base || url

      if (options.processingMode) {
        context.DEFAULT_PROCESSING_MODE = options.processingMode
      }

      let ordered = true

      let sourceData = loadJson(src)

      let contextData
      if (tc.context) {
        let contextSrc = TESTSUITE_DIR + tc.context
        contextData = loadJson(contextSrc)
        sourceData = expand(sourceData, baseUri, null, ordered)
      }

      let expectedSrc = TESTSUITE_DIR + tc.expect
      let expectedData = loadJson(expectedSrc)
      if (CONTEXT in expectedData) {
        delete expectedData[CONTEXT]
      }

      let resultData

      let expandContext = null
      if (options.expandContext) {
          expandContext = TESTSUITE_DIR + options.expandContext
      }
      resultData = expand(sourceData, baseUri, expandContext, ordered)

      if (tc[TYPE].indexOf('jld:FlattenTest') > -1) {
        try {
          resultData = flatten(resultData, ordered)
        } catch (e) {
          t.fail(e)
        }
      }

      if (contextData) {
        let compactArrays = options.compactArrays
        if (compactArrays === void(0)) compactArrays = true
        try {
          resultData = compact(contextData, sourceData, baseUri, compactArrays, ordered)
        } catch (e) {
          t.fail(e)
        }
      }

      t.deepEqual(resultData, expectedData)
      //t.equal(JSON.stringify(resultData, null, 2), JSON.stringify(expectedData, null, 2))
      t.end()
    })
  }
}

testCaseRunner('expand')
testCaseRunner('compact')
testCaseRunner('flatten')
